### PR TITLE
fix: trim whitespace from auth token (#10539)

### LINKED ISSUE #10539: [CLI] Misleading error message with whitespace in --auth-token
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

If the argument to --auth-token mistakenly contains whitespace (e.g. a newline accidentally captured when base64-encoding a secret), the following misleading error will be produced:

```
time="2022-09-06T22:19:04Z" level=fatal msg="rpc error: code = Internal desc = stream terminated by RST_STREAM with error code: PROTOCOL_ERROR"
```

for anyone who is not well-versed in gRPC internals, this is not easy to debug.

**To Reproduce**

```bash
export TOKEN='<some valid token>
'
argocd --auth-token="${TOKEN}" --server=argocd.example.com version
```

**Expected behavior**

Validation of the allowable characters in the token, or at least an entry be added to the FAQ for this error.

**Screenshots**

N/A

**Version**

```shell
argocd: v2.3.2+ecc2af9
  BuildDate: 2022-03-23T00:40:57Z
  GitCommit: ecc2af9dcaa12975e654cde8cbbeaffbb315f75c
  GitTreeState: clean
  GoVersion: go1.17.6
  Compiler: gc
  Platform: linux/amd64
argocd-server: v2.3.2+ecc2af9
  BuildDate: 2022-03-23T00:40:57Z
  GitCommit: ecc2af9dcaa12975e654cde8cbbeaffbb315f75c
  GitTreeState: clean
  GoVersion: go1.17.6
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: v4.4.1 2021-11-11T23:36:27Z
  Helm Version: v3.8.0+gd141386
  Kubectl Version: v0.23.1
  Jsonnet Version: v0.18.0
```

**Logs**

See above

### PR DESCRIPTION
**Pull request by cleverhu**
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
Fixes: https://github.com/argoproj/argo-cd/issues/10539

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by cleverhu** [comment]
@crenshaw-dev PTAL

File: pkg/apiclient/apiclient.go [hunk 1]
```diff
@@ -234,6 +234,12 @@ func NewClient(opts *ClientOptions) (Client, error) {
 	}
 	if opts.AuthToken != "" {
 		c.AuthToken = opts.AuthToken
+		parser := jwt.NewParser(jwt.WithoutClaimsValidation())
+		claims := jwt.RegisteredClaims{}
+		_, _, err = parser.ParseUnverified(c.AuthToken, &claims)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
```
Comment (ashutosh16) [inline pkg/apiclient/apiclient.go:241]: format the error before returning the error to the user's `fmt.Errorf("error doing something: %v", err)`

Comment (cleverhu) [reply pkg/apiclient/apiclient.go]: I import the func from other util package, is that ok?

**Review by ashutosh16 — APPROVED** [review]
LGTM

File: pkg/apiclient/apiclient.go [hunk 2]
```diff
@@ -234,6 +235,10 @@ func NewClient(opts *ClientOptions) (Client, error) {
 	}
 	if opts.AuthToken != "" {
 		c.AuthToken = opts.AuthToken
+		_, err := clusterauth.ParseServiceAccountToken(c.AuthToken)
```
Comment (crenshaw-dev) [inline pkg/apiclient/apiclient.go:238]: This function looks specifically designed to parse k8s ServiceAccount tokens. I'm not sure whether it's appropriate for validating Argo CD auth tokens.

What if we just added logic to the CLI to strip leading and trailing whitespace from the token? That seems reasonably safe and would solve the reported issue.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: pkg/apiclient/apiclient.go (hunk 2 above)
Comment (cleverhu) [reply pkg/apiclient/apiclient.go:238]:
```
	if localCfg != nil {
		err = c.refreshAuthToken(localCfg, ctxName, opts.ConfigPath)
		if err != nil {
			return nil, err
		}
	}
```

```
	if err != nil {
		return err
	}
	parser := jwt.NewParser(jwt.WithoutClaimsValidation())
	var claims jwt.RegisteredClaims
	_, _, err = parser.ParseUnverified(configCtx.User.AuthToken, &claims)
	if err != nil {
		return err
	}
	if claims.Valid() == nil {
		// token is still valid
		return nil
	}
```
I think the validating for tokens is the same.

Comment (cleverhu) [reply pkg/apiclient/apiclient.go:238]: I copied the `refreshAuthToken` code when I submitted it for the first time, but it was not elegant.So I imported methods from other packages.

Comment (crenshaw-dev) [reply pkg/apiclient/apiclient.go:238]: @cleverhu honestly, I'd like the space-trimming option much better. It solves the reported issue, and has a very very low risk of accidentally breaking something.

Comment (cleverhu) [reply pkg/apiclient/apiclient.go:238]: Ok，I will change the code.

Comment (cleverhu) [reply pkg/apiclient/apiclient.go]: @crenshaw-dev I verified that this is valid, and updated the code.PTAL.

**Comment by cleverhu** [comment]
@crenshaw-dev  I verified that this is valid, and updated the code.PTAL.

**Review by crenshaw-dev — APPROVED** [review]
Local test looks good!

Not gonna cherry-pick back, because it touches a pretty important piece of code, and it only solves an annoyance.

**Pull request merged**